### PR TITLE
Desy Bump Bonding test working now for module

### PR DESCRIPTION
**Pull request by andreavargas**
Changes
- calling now getEfficiencyVsDAC routine for the complete module
- getting working point in tornado plot for CALS settings 
- used time should be ~200 sec for the complete module ;)

Please modify the "testParameters.dat" file with

-- BB2
targetIa           28
PIX                 2,2
Ntrig               10
vcals               250
Plwidth            35

### CODE REVIEW

**Comment by ursl** [comment]
Hi Andrea, 

I have a few questions: 

1) Why do redo setVana() in the BB test?

2) What is the reason to allow for a non-standard targetIa?

3) What are the differences if you do NOT do setVthrCompCalDelForCals()?  I understand that you could argue that it is "better" to redo it, but I'd like to know by how much better it is. Do you have quantitative studies on this? 

For both cases, I am  not really happy that you duplicate functionality that is already in pxar into this specific test. 

Cheers,
--U.

**Comment by andreavargas** [comment]
Dear Urs,

thank you for having a look into the routine,

For 1) & 2)
yes our bb-test needs the optimization of the "tornado" distribution for 
*cals* and *ctrlreg 4*. If I undertood correctly, in the pretest code, 
the optimization takes place for ctrlreg 0 and not for cals. Since our 
bonding bumps are bigger than the psi ones, our signal is smaller. We 
can increase the signal by allowing a non standard analog current of 28 mA.

For 3)
I check our code without optimization of the tornado plot  and I got 
many missing O(150) bumps, which actually are fine. That was reason we 
decided to include those routines in the test, they do not take much 
time - something like 5sec or less.

Please let me know your opinion,
Cheers,

Andrea

**Comment by ursl** [comment]
Hi Andrea, 

many thanks. OK, I guess then you should keep your two routines. 

But I think your current setup puts the DUT into a state which is not the same as before the BB2 test. cacheDacs() is not recursive (and there are no plans to make it work so). You do call it in doTest(), you then also call it in setVana() and setVthrCompCalDelForCals(). I am pretty sure that after the BB2 test the DUT is configured differently than before. That should not be the case. 

I think the best would be to use in PixTestBB2Map::doTest() from PixTest

std::vector<uint8_t> getDacs(std::string dacName); 
void setDacs(std::string dacName, std::vector<uint8_t> dacVector); 

for those DACs (ctrlreg, vcal, vthrcomp, caldel, ...) that you do modify, and at the end restore them to their original values. It might also be required to have another look at your usage of cacheDacs() in your subroutines for PixTestBB2Map (I have not done this so far). 

The best way to be sure that you leave the DUT in a "safe" (i.e. unchanged) state would be to save the DAC files before the BB2 test, (move them somewhere), run the BB2 test, save the DAC files, and compare to the original ones. 

Sorry I missed this point in my first look at your code. 

Cheers,
--U.

**Comment by andreavargas** [comment]
Dear Urs,

yes thank you, I will follow your suggestions and check that the DUT is 
leaved in a safe unchanged mode. I will
let you know as soon as possible.

Cheers,
Andrea

**Comment by andreavargas** [comment]
Dear Urs,

I made the changes to the BB2 routine. Now the DAC values are unchanged 
after running the test. I commited and push the changes to my fork, but  
I am not sure if the changes are also now in the pull request?

Please let me know your opinion!
Best Wishes,

Andrea

File: tests/PixTestBB2Map.cc [hunk 1]
```diff
@@ -262,4 +357,344 @@ void PixTestBB2Map::doTest() {
...
+  
+  double i016 = fApi->getTBia()*1E3;
+
+  // FIXME this should not be a stopwatch, but a delay
+  TStopwatch sw;
+  sw.Start(kTRUE); // reset
+  do {
+    sw.Start(kFALSE); // continue
+    i016 = fApi->getTBia()*1E3;
+  } while (sw.RealTime() < 0.1);
```
Comment (simonspa) [inline tests/PixTestBB2Map.cc]: Were you looking for something like `pxar::mDelay(uint32_t milliseconds)`?
https://github.com/psi46/pxar/blob/master/core/utils/helper.h#L40

**Pull request merged**